API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#ifndef HERMES_ABI_HERMES_ABI_H
+#define HERMES_ABI_HERMES_ABI_H
+
+#include <assert.h>
+#include <stdalign.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#ifndef HERMES_EXPORT
+#ifdef _MSC_VER
+#define HERMES_EXPORT __declspec(dllexport)
+#else // _MSC_VER
+#define HERMES_EXPORT __attribute__((visibility("default")))
+#endif // _MSC_VER
+#endif // !defined(HERMES_EXPORT)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// Define the structure for references to pointer types in JS (e.g. string,
+/// object, BigInt).
+/// TODO: Replace jsi::PointerValue itself with this C implementation to
+/// eliminate pointer management overhead in the JSI wrapper.
+struct HermesABIManagedPointer;
+struct HermesABIManagedPointerVTable {
+  /// Pointer to the function that should be invoked when this reference is
+  /// released.
+  void (*invalidate)(HermesABIManagedPointer *);
```

Comment:
Please add explicit calling conventions to all functions and function pointers.
We technically need it only for MSVC x86.
The best way to do it is to define a macro, and then use it everywhere. E.g. see the `NAPI_CDECL` in [Node-API](https://github.com/nodejs/node/blob/main/src/js_native_api_types.h).

Reply:
The function list here definitely needs a lot of work (documentation, parameter names, calling conventions)